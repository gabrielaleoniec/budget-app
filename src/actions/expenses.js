import database from '../firebase/firebase';

// ADD EXPENSE
export const addExpense = (expense = {}) => {
    return {
        type: "ADD_EXPENSE",
        expense
    }
}

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const date = createdAt ? new Date(createdAt).valueOf() : new Date().valueOf();

        const expense = { description, note, amount, createdAt: date };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                ...expense,
                id: ref.key
            }))
        });
    }
}

// REMOVE EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: "REMOVE_EXPENSE",
    id: id
});

// START REMOVE EXPENSES
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => { 
            dispatch(removeExpense({ id }))
        });
    }
}

// EDIT EXPENSE
export const updateExpense = (id, updates) => ({
    type: "UPDATE_EXPENSE",
    id,
    updates
})

// SET EXPENSES
export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then(snapshot => {
            const expenses = [];

            snapshot.forEach(childSnapshot => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        }).catch((e) => {
            console.log('Error', e);
        })
    }
};
