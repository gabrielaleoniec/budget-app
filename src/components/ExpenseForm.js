import React from 'react';
import moment from 'moment';

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt).format('YYYY-MM-DD') : moment().format('YYYY-MM-DD'),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        this.setState(
            {
                description: e.target.value
            }
        )
    }

    onNoteChange = (e) => {
        this.setState({
            note: e.target.value
        })
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (amount.match(/^\d*(\.\d{1,2})?$/)) {
            this.setState({
                amount: e.target.value
            })
        }
    }

    onCreatedAtChange = (e) => {
        this.setState({
            createdAt: e.target.value
        })
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            // Error
            return this.setState(() => ({
                error: 'Please provide description and amount'
            }))
        }

        // Clear the error
        this.setState(() => ({
            error: ''
        }));

        this.props.onSubmit({
            description: this.state.description,
            note: this.state.note,
            amount: parseFloat(this.state.amount, 10) * 100,
            createdAt: this.state.createdAt
        });
    }

    render() {
        return (
            <form onSubmit={this.onSubmit} className="form input-group">
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <label className="input-group__item">
                <span className="sr-only">Description</span>
                    <input
                        type="text"
                        name="description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        className="text-input"
                        placeholder="Description" />
                </label>
                <label className="input-group__item">
                <span className="sr-only">Note</span>
                    <textarea
                        name="note"
                        onChange={this.onNoteChange}
                        placeholder="Put some note here"
                        value={this.state.note}
                        className="textarea"
                    >
                    </textarea>
                </label>
                <label className="input-group__item">
                <span className="sr-only">Amount</span>
                    <input
                        type="number"
                        name="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        className="text-input"
                        placeholder="Amount"
                    />
                </label>
                <label className="input-group__item">
                <span className="sr-only">Created</span>
                    <input
                        type="date"
                        name="createdAt"
                        value={this.state.createdAt}
                        onChange={this.onCreatedAtChange}
                        className="select" />
                </label>
                <button type="submit" className="button">Save expense</button>
            </form>
        );
    }
}

export default ExpenseForm;