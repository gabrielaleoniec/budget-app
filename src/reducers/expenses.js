const defaultStateExpensifyReducer = [];

const expensifyReducer = (state = defaultStateExpensifyReducer, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'UPDATE_EXPENSE':
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                return expense;
            })
        case 'SET_EXPENSES':
            return action.expenses
        default:
            return state;
    }
}

export default expensifyReducer;