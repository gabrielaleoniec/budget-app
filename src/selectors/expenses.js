
const getVisibleExpenses = (expenses, { text, sortBy, sortByAsc, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const matchText = expense.description.toLowerCase().includes(text.toLowerCase());
        const matchStartDate = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const matchEndDate = typeof endDate !== 'number' || expense.createdAt <= endDate;

        return matchText && matchStartDate && matchEndDate;
    }).sort((a, b) => {
        if (sortBy === 'createdAt') {
            return sortByAsc ? b.createdAt - a.createdAt : a.createdAt - b.createdAt;
        } else if (sortBy === 'amount') {
            return sortByAsc ? b.amount - a.amount : a.amount - b.amount;
        }
    })
}

export default getVisibleExpenses;