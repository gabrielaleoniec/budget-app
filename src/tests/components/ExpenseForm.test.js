import React from 'react';
import { shallow } from 'enzyme';
import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../fixtures/expenses';
import moment from 'moment';

test('Should render ExpenseForm correctly', () => {
    const wrapper = shallow(<ExpenseForm />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseForm correctly with expense data', () => {
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render error when required data not provided', () => {
    const wrapper = shallow(<ExpenseForm />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test('Should set description on input change', () => {
    const value = 'Bicycle repair';
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} />)
    wrapper.find('input').at(0).simulate('change', {
        target: {
            value
        }
    });
    expect(wrapper.state('description')).toBe(value);
    expect(wrapper).toMatchSnapshot();
});

test('Should set note on textarea change', () => {
    const value = 'Better save';
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} />);
    wrapper.find('textarea').simulate('change', {
        target: {
            value
        }
    });
    expect(wrapper.state('note')).toBe(value);
    expect(wrapper).toMatchSnapshot();
});

test('Should set amount on input change', () => {
    const value = '33300';
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} />);
    wrapper.find('input').at(1).simulate('change', {
        target: {
            value
        }
    });
    expect(wrapper.state('amount')).toBe(value);
    expect(wrapper).toMatchSnapshot();
});

test('Should not set amount on input change when value is not numeric', () => {
    const value = '12.222';
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} />);
    wrapper.find('input').at(1).simulate('change', {
        target: {
            value
        }
    });
    expect(wrapper.state('amount')).toBe((expenses[0].amount / 100).toString());
    expect(wrapper).toMatchSnapshot();
});

test('Should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('error')).toBe('');
    const createdAt = moment(expenses[0].createdAt).format('YYYY-MM-DD');
    // const createdAt = date.getFullYear() + '-' + (date.getMonth() + 1).toString().padStart(2, '0') + '-' + date.getDate();
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        description: expenses[0].description,
        note: expenses[0].note,
        amount: expenses[0].amount,
        createdAt
    });
});

test('Should set the expense date in input change', () => {
    const value = '2020-11-15';
    const wrapper = shallow(<ExpenseForm expense={expenses[0]}/>);

    wrapper.find('input').at(2).simulate('change', {
        target: {
            value
        }
    });

    expect(wrapper.state('createdAt')).toBe(value);
})