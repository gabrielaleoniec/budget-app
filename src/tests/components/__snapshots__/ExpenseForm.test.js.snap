// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should not set amount on input change when value is not numeric 1`] = `
<form
  className="form input-group"
  onSubmit={[Function]}
>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Description
    </span>
    <input
      autoFocus={true}
      className="text-input"
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Shoes"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Note
    </span>
    <textarea
      className="textarea"
      name="note"
      onChange={[Function]}
      placeholder="Put some note here"
      value="Too expensive"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Amount
    </span>
    <input
      className="text-input"
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value="450"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Created
    </span>
    <input
      className="select"
      name="createdAt"
      onChange={[Function]}
      type="date"
      value="2020-06-14"
    />
  </label>
  <button
    className="button"
    type="submit"
  >
    Save expense
  </button>
</form>
`;

exports[`Should render ExpenseForm correctly 1`] = `
<form
  className="form input-group"
  onSubmit={[Function]}
>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Description
    </span>
    <input
      autoFocus={true}
      className="text-input"
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Note
    </span>
    <textarea
      className="textarea"
      name="note"
      onChange={[Function]}
      placeholder="Put some note here"
      value=""
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Amount
    </span>
    <input
      className="text-input"
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value=""
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Created
    </span>
    <input
      className="select"
      name="createdAt"
      onChange={[Function]}
      type="date"
      value="1970-01-01"
    />
  </label>
  <button
    className="button"
    type="submit"
  >
    Save expense
  </button>
</form>
`;

exports[`Should render ExpenseForm correctly with expense data 1`] = `
<form
  className="form input-group"
  onSubmit={[Function]}
>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Description
    </span>
    <input
      autoFocus={true}
      className="text-input"
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Shoes"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Note
    </span>
    <textarea
      className="textarea"
      name="note"
      onChange={[Function]}
      placeholder="Put some note here"
      value="Too expensive"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Amount
    </span>
    <input
      className="text-input"
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value="450"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Created
    </span>
    <input
      className="select"
      name="createdAt"
      onChange={[Function]}
      type="date"
      value="2020-06-14"
    />
  </label>
  <button
    className="button"
    type="submit"
  >
    Save expense
  </button>
</form>
`;

exports[`Should render error when required data not provided 1`] = `
<form
  className="form input-group"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Please provide description and amount
  </p>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Description
    </span>
    <input
      autoFocus={true}
      className="text-input"
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Note
    </span>
    <textarea
      className="textarea"
      name="note"
      onChange={[Function]}
      placeholder="Put some note here"
      value=""
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Amount
    </span>
    <input
      className="text-input"
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value=""
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Created
    </span>
    <input
      className="select"
      name="createdAt"
      onChange={[Function]}
      type="date"
      value="1970-01-01"
    />
  </label>
  <button
    className="button"
    type="submit"
  >
    Save expense
  </button>
</form>
`;

exports[`Should set amount on input change 1`] = `
<form
  className="form input-group"
  onSubmit={[Function]}
>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Description
    </span>
    <input
      autoFocus={true}
      className="text-input"
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Shoes"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Note
    </span>
    <textarea
      className="textarea"
      name="note"
      onChange={[Function]}
      placeholder="Put some note here"
      value="Too expensive"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Amount
    </span>
    <input
      className="text-input"
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value="33300"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Created
    </span>
    <input
      className="select"
      name="createdAt"
      onChange={[Function]}
      type="date"
      value="2020-06-14"
    />
  </label>
  <button
    className="button"
    type="submit"
  >
    Save expense
  </button>
</form>
`;

exports[`Should set description on input change 1`] = `
<form
  className="form input-group"
  onSubmit={[Function]}
>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Description
    </span>
    <input
      autoFocus={true}
      className="text-input"
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Bicycle repair"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Note
    </span>
    <textarea
      className="textarea"
      name="note"
      onChange={[Function]}
      placeholder="Put some note here"
      value="Too expensive"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Amount
    </span>
    <input
      className="text-input"
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value="450"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Created
    </span>
    <input
      className="select"
      name="createdAt"
      onChange={[Function]}
      type="date"
      value="2020-06-14"
    />
  </label>
  <button
    className="button"
    type="submit"
  >
    Save expense
  </button>
</form>
`;

exports[`Should set note on textarea change 1`] = `
<form
  className="form input-group"
  onSubmit={[Function]}
>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Description
    </span>
    <input
      autoFocus={true}
      className="text-input"
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value="Shoes"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Note
    </span>
    <textarea
      className="textarea"
      name="note"
      onChange={[Function]}
      placeholder="Put some note here"
      value="Better save"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Amount
    </span>
    <input
      className="text-input"
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="number"
      value="450"
    />
  </label>
  <label
    className="input-group__item"
  >
    <span
      className="sr-only"
    >
      Created
    </span>
    <input
      className="select"
      name="createdAt"
      onChange={[Function]}
      type="date"
      value="2020-06-14"
    />
  </label>
  <button
    className="button"
    type="submit"
  >
    Save expense
  </button>
</form>
`;
